PHP/PHP Source.sublime-syntax

```diff
@@ -1,698 +1,698 @@
 %YAML 1.2
+# https://www.php.net/manual/en/langref.php
 ---
 name: PHP Source
 scope: source.php
 hidden: true
 version: 2
 
 variables:
+  guarded_identifier: (?!{{reserved}}){{identifier}}
+  identifier: '{{identifier_start}}[[:alnum:]_]*'
   identifier_start: '[[:alpha:]_]'
-  identifier: '\b{{identifier_start}}[[:alnum:]_]*\b'
-  path: '\\?({{identifier}}\\)*{{identifier}}'
+
   regex_modifier: '[eimsuxADJSUX]*'
-  sql_indicator: \s*(?:SELECT|INSERT|UPDATE|DELETE|CREATE|REPLACE|ALTER)(?=\s)
-  visibility_modifier: '\b(?:var|public|private|protected)\b'
+  sql_indicator: |-
+    (?x: \s* (?:
+    # dml statements
+      SELECT | INSERT | REPLACE | DELETE | TRUNCATE | UPDATE
+    # ddl statements
+    | ADD | ALTER | CREATE | DROP
+    # conditional
+    | IF \s+ (?: NOT \s+ )? EXISTS
+    # declaration
+    | DECLARE | WITH
+    ) \s )
 
   # A pattern which terminates an incomplete expression at any point.
-  # `\s*\?>` is for fixing one-liner ( https://github.com/sublimehq/Packages/issues/1545 )
-  expression_end: (?=;|}|\s*\?>|{{keywords}})
+  # `\?>` is for fixing one-liner ( https://github.com/sublimehq/Packages/issues/1545 )
+  expression_end: (?=;|}|\?>|{{keywords}})
+
+  heredoc_quoted_end: (?=^\s*\4\b)
+  heredoc_unquoted_end: (?=^\s*\2\b)
+
+  # no odd number of backslashes behind (`\` or `\\\`)
+  # note:
+  #  1. For use in `escape` commands only !!
+  #  2. ST's Regex Compatibility Check marks it as incompatible
+  #     as the real usage of the variable is not checked.
+  #     It's a false positive as long as (1.) is respected.
+  no_escape_behind: (?<![^\\]\\)(?<![^\\][\\]{3})
+
+  reserved: |-
+    (?xi:
+      {{keywords}}
+    | {{builtin_functions}}
+    | {{include_directives}}
+    | {{logical_operators}}
+    | {{storage_keywords}}
+    | {{storage_modifiers}}
+    | {{storage_types}}
+    | {{class_modifier}}
+    | {{builtin_variables}}
+    | function\b
+    | as\b
+    )
 
   keywords: |-
-    (?x:
-      {{keywords_control}}
-    | {{keywords_flow}}
-    | {{keywords_declaration}}
-    | {{visibility_modifier}}
+    (?xi:
+      {{control_keywords}}
+    | {{flow_keywords}}
+    | {{declaration_keywords}}
+    | {{access_modifier}}
+    | {{namespace_modifier}}
     )
 
-  keywords_declaration: |-
+  # Statement Level Keywords
+  declaration_keywords: |-
+    (?xi: class | const | declare | enddeclare | interface | namespace | trait | use | var )\b
+  control_keywords: |-
+    (?xi: case | catch | default | do | (?: end )? (?: for(?:each)? | if | switch | while )
+    | else | elseif | finally | switch | try )\b
```

Comment:
and `enum` for `declaration_keywords` btw.

Reply:
All `statement level keywords` pop `expressions` context off stack (terminate it).

1. `match` is used like a function-call which returns a value and therefore looks like being valid in the middle of an expression. Adding it to `control_keywords` would cause any expression to be popped off stack, which seems wrong with the way it works.

```
return match { ... }
```

also valid

```
$var = "true";

if ( match($var) { "true" => true, "false" => false } ) { echo "match returned true"; } else { echo "or no "; }
```

3. `enum` is not a reserved keyword (according to https://www.php.net/manual/en/reserved.keywords.php). PHP treats it as ordinary constant in expressions. Hence `enum` also shouldn't pop the `expressions` syntax off stack, too.

Reply:
Gotcha, makes sense! Like I said, way over my head 😂

Reply:
Using the syntax in everyday projects and reporting back any kinds of annoyances etc. would be appriciated.